API key enhancements

itisAliRH opened the pull request:
Closes #13328 and Closes #10742

This pr migrates the API key endpoint to Fast API, vuetify the client, and the user can mark the API key as deleted.

- [x] Client
- - [x] List the latest created key
- - [x] Create a new key when there is no key available
- - [x] Delete a key
- - [x] Add route
- - [x] Add to user preferences list
- - [x] Delete old UI files
- [x] Backend
- - [x] Create managers
- - [x] Create services
- - [x] Add endpoints
- - [x] Delete old endpoints

## API changes
We made some retro-compatible API changes around the API key management endpoints:
- The `APIKeys` model now has a `deleted` column. It has been added to the galaxy model and the toolshed model.
- **Dropped** the `GET /api/users/{id}/api_key/inputs` that returned additional data for rendering the API key in the UI.
- **Dropped** the `PUT /api/users/{id}/api_key/inputs` that created a new API key and returned data for rendering the UI.
- **Added** the `GET /api/users/{id}/api_key/detailed` that now returns the `APIKeyModel` or 204 status code if there's no key.
  ```python
  class APIKeyModel(BaseModel):
      key: str = Field(Required, title="Key", description="API key to interact with the Galaxy API")
      create_time: datetime = Field(
          Required, title="Create Time", description="The time and date this API key was created."
      )
  ```
- **Added** `DELETE /api/users/{user_id}/api_key` to delete/revoke (all*) current API key.
- The rest of the related endpoints preserve their usual behaviour.


**Important Note (*)**: Since the old button to generate an API was always enabled, a user might have multiple API keys in the database if it was used multiple times. Deleting the API key will mark all existing API keys as deleted.


## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

itisAliRH committed:
Add new api_keys managers: `get_api_keys` and `delete_api_key`

itisAliRH committed:
Add API key services in users services

itisAliRH committed:
Add new API keys routes in users api

itisAliRH committed:
Create APIKeyItem.vue

itisAliRH committed:
Create APIKey.vue

itisAliRH committed:
Create APIKey services

itisAliRH committed:
Add new APIKey routes and user preferences menu

itisAliRH committed:
Add trash icon to api key button

itisAliRH committed:
Limit to display only last created API key

itisAliRH committed:
Return only last created API key

itisAliRH committed:
Delete old API key end points

itisAliRH committed:
Delete old api_key methods, use the old paths and add types

itisAliRH committed:
Fix create_api_key type error

itisAliRH committed:
Add types to api_keys manager

itisAliRH committed:
Fix selenium navigation get_api_key paths

itisAliRH committed:
Add todo to return all api_keys when ready

itisAliRH committed:
Fix create api key test

itisAliRH committed:
Fix client api call paths and ui improvments

itisAliRH committed:
Fix accessibility issue

itisAliRH committed:
python lint and format

davelopez committed:
Remove unnecessary query param in API tests teardown

davelopez committed:
Fix user handling in api_key endpoints

davelopez committed:
Add pydantic model and refactor API documentation

davelopez committed:
Adapt UI to API changes

The UI will assume multiple keys for future scope keys, but the backend will return just one for now

davelopez committed:
Handle no content response in Selenium tests

Adapt get_api_key function to return

davelopez committed:
Adapt test_api_key selenium test to new UI

davelopez committed:
Refactor remove redundant code in UsesApiTestCaseMixin

`ensure_user_with_email` already returns the user with the corresponding email

davelopez committed:
Add API tests coverage for API key management

davelopez committed:
Fix exception not being properly raised

itisAliRH committed:
Change API key item UI and a add title

mvdbeek commented on lib/galaxy/managers/api_keys.py:
```diff
@@ -1,26 +1,40 @@
 
-    def create_api_key(self, user) -> str:
+    def get_api_keys(self, user: User) -> List[APIKeys]:
+        return user.api_keys
```
That is so weird, users should only have a single active API key.

davelopez replied:
Now the inactive ones will be marked as deleted :)

mvdbeek commented on lib/galaxy/managers/api_keys.py:
```diff
@@ -1,26 +1,40 @@
         return key
+
+    def delete_api_key(self, user: User, key: str) -> None:
+        sa_session = self.app.model.context
```
Maybe mark all keys as deleted ?

davelopez replied:
Done in [9e10e97](https://github.com/galaxyproject/galaxy/pull/14489/commits/9e10e97b9cbf88c13c05884d6ce6bf69b99d9f81) :)

mvdbeek reviewed (commented):
> So removing one API key will obviously just show the next available one.

That's problematic, I suggest we mark all old keys as deleted.

I would not change the return value for API keys, I don't think we'll introduce "scoped"  api keys, this should use a different infrastructure for generating tokens, with a different endpoint.

davelopez committed:
Add deleted column to API keys table

davelopez committed:
Restore old API behavior and add detailed endpoint

The existing endpoints will return the API key as a plain string and the `GET /api/users/{user_id}/api_key` route will get or create the key as it used to be.

davelopez committed:
Refactor delete API key to mark as deleted all keys

This will make sure any user with multiple created keys will mark them as deleted as soon as the first one is deleted.

davelopez committed:
Adapt UI to API changes

davelopez committed:
Make sure we return always the last created API key

davelopez committed:
Add unit test for api_keys.deleted column

davelopez committed:
Use APIKeys model from app

Galaxy and the ToolShed have duplicated APIKeys models and the ApiKeyManager can be used from both apps

davelopez committed:
Add deleted column to api_keys table in ToolShed

Since the toolshed uses a duplicated `api_keys` table we sadly need to duplicate the migration too...

davelopez committed:
Cleanup unused code and mixin

itisAliRH committed:
Merge branch 'dev' into api-key-Enhancement

itisAliRH committed:
update APIKeyItem with composition API and improve UI

itisAliRH committed:
update APIKey with composition API
